### PR TITLE
Enable SSE2 compression path to work on MSVC

### PR DESCRIPTION
**Pull request by TrianglesPCT**
The compile time detection uses the `__SSE2__ `predefined macro, but this does not exist on MSVC--resulting it that compiler always using scalar path.

On MSVC you can detect SSE support by checking for `_M_AMD64`, as when targeting x64 it requires SSE2 as a baseline, this patch adds this check.

### CODE REVIEW

**Review by TrianglesPCT — COMMENTED** [review]
.

**Review by TrianglesPCT — COMMENTED** [review]
Switch to unaligned load

**Comment by Cyan4973** [comment]
Thanks @TrianglesPCT ,

did you attempt to measure performance differences provided by your PR ?

File: lib/compress/zstd_lazy.c [hunk 1]
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
...
+	ZSTD_Vec256 v;
+    v.v = _mm256_set1_epi8((char)val);
+	return v;
+}
+
+static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
+    return (ZSTD_VecMask)_mm256_movemask_epi8(_mm256_cmpeq_epi8(x.v, y.v));
+}
+
+#endif
```
Comment (terrelln) [inline lib/compress/zstd_lazy.c]: Can you submit this portion as a separate PR please?

It should be good go go with some minor formatting changes, but this portion requires performance testing, where the MSVC fix is easy to accept.

I know I measured the performance of AVX2 vs SSE, and it was slightly better. But, I didn't see a huge speed-up.

File: lib/compress/zstd_lazy.c [hunk 2]
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
...
+
+static ZSTD_Vec256 ZSTD_Vec256_read(const void* const ptr) {
+	ZSTD_Vec256 v;
+    v.v = _mm256_loadu_si256((const __m256i*)ptr);
+	return v;
+}
+
+static ZSTD_Vec256 ZSTD_Vec256_set8(BYTE val) {
+	ZSTD_Vec256 v;
+    v.v = _mm256_set1_epi8((char)val);
```
Comment (terrelln) [inline lib/compress/zstd_lazy.c]: nit: formatting

File: lib/compress/zstd_lazy.c [hunk 3]
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
   sndMask = ZSTD_Vec128_cmpMask8(x.snd, y.snd);
   return fstMask | (sndMask << 16);
 }
+#else//AVX2
```
Comment (terrelln) [inline lib/compress/zstd_lazy.c]: nit: We are C90, so you need to do `/* AVX2 */`

File: lib/compress/zstd_lazy.c (hunk 1 above)
Comment (TrianglesPCT) [reply lib/compress/zstd_lazy.c]: Yea I'll remove this and submit it seperate

**Comment by TrianglesPCT** [comment]
> 
> 
> Thanks @TrianglesPCT ,
> 
> did you attempt to measure performance differences provided by your PR ?

I removed the AVX2 stuff for a later PR

**Comment by Cyan4973** [comment]
Well, even without `AVX2`,
this PR is presumed to help performance, because it's supposed to help `SSE2` detection ?

**Comment by TrianglesPCT** [comment]
All it is doing is normalizing MSVC to target the same code path as GCC/Clang when SSE2 is set as a requirement for your compile target. Not adding any new code..

 The `defined(__SSE2__)` check doesn't work on MSVC because that define doesn't exist even if SSE2 is enabled.

 If you want specific performance comparison I can do that might be few days until I have time for that.

**Comment by terrelln** [comment]
> If you want specific performance comparison I can do that might be few days until I have time for that.

That would be great! But that doesn't need to block merging this PR. You can easily measure performance with the builtin benchmark tool.

```
zstd -b5e12 silesia.tar
```

**Comment by ghost** [comment]
Will there be a hot fix for v1.5.0?

**Comment by wolfpld** [comment]
> You can easily measure performance with the builtin benchmark tool.

The difference is rather spectacular. Results for 5950X.

Before:
```
 5#silesia.tar       : 211957760 ->  63806593 (3.322), 111.0 MB/s , 977.5 MB/s
 6#silesia.tar       : 211957760 ->  62980544 (3.365), 107.9 MB/s ,1001.6 MB/s
 7#silesia.tar       : 211957760 ->  61482185 (3.447),  73.1 MB/s ,1055.9 MB/s
 8#silesia.tar       : 211957760 ->  60914308 (3.480),  57.2 MB/s ,1076.1 MB/s
 9#silesia.tar       : 211957760 ->  59928587 (3.537),  52.1 MB/s ,1100.2 MB/s
10#silesia.tar       : 211957760 ->  59296205 (3.575),  50.3 MB/s ,1103.6 MB/s
11#silesia.tar       : 211957760 ->  59152928 (3.583),  47.7 MB/s ,1106.8 MB/s
12#silesia.tar       : 211957760 ->  58640204 (3.615),  30.6 MB/s ,1123.8 MB/s
```

After:
```
 5#silesia.tar       : 211957760 ->  63806593 (3.322), 160.8 MB/s , 980.0 MB/s
 6#silesia.tar       : 211957760 ->  62980544 (3.365), 153.3 MB/s ,1004.7 MB/s
 7#silesia.tar       : 211957760 ->  61482185 (3.447), 111.9 MB/s ,1056.3 MB/s
 8#silesia.tar       : 211957760 ->  60914308 (3.480),  90.9 MB/s ,1079.5 MB/s
 9#silesia.tar       : 211957760 ->  59928587 (3.537),  74.7 MB/s ,1101.9 MB/s
10#silesia.tar       : 211957760 ->  59296205 (3.575),  71.3 MB/s ,1107.0 MB/s
11#silesia.tar       : 211957760 ->  59152928 (3.583),  65.0 MB/s ,1113.2 MB/s
12#silesia.tar       : 211957760 ->  58640204 (3.615),  51.3 MB/s ,1129.6 MB/s
```

**Review by Cyan4973 — APPROVED** [review]

**Review by senhuang42 — APPROVED** [review]
This is a great fix, thanks for the contribution!
